### ISSUE TITLE
Installation of h5 package on Mac OS X

### ISSUE THREAD

**Issue by pcarbo** [mannau/h5#44]
Thank you for the nice package. 

It seems that the installation steps need to be clarified and improved on. For example, I tried to use the configure script on my computer (Mac) but it didn't seem to work. So I resorted to this install command, 

```bash
R CMD INSTALL h5 --configure-vars="'LIBS=-L/usr/local/Cellar/hdf5/1.10.0-patch1 -L. -lhdf5_cpp -lhdf5 -lz -lm' CPPFLAGS='-I/usr/local/include'"
```

which worked, but it required that I set all the libraries manually, and it should be done internally via `src/Makevars`, and allow me to set additional flags (e.g., paths to libraries) via `~/.R/Makevars` or by passing options to `R CMD INSTALL`. For more details see [here](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Using-Makevars) and [here](https://cran.r-project.org/doc/manuals/r-release/R-admin.html#Customizing-package-compilation).

Note that the `R CMD INSTALL` command you suggest in the README includes a bunch of additional libraries that don't seem necessary, and addition you got the quotes wrong.

Also, I was able to install with the hdf5 libraries in Anaconda 3:

```bash
R CMD INSTALL h5 --configure-vars="LIBS='-L/Users/pcarbo/anaconda3/lib -L. -lhdf5_cpp -lhdf5 -lz -lm' CPPFLAGS='-I/usr/local/include'"
```

i hope all these suggestions are useful.

**Comment by mannau** [comment]
Hi Peter,

Am 03/03/17 um 19:07 schrieb Peter Carbonetto:
> Thank you for the nice package.
> 
> It seems that the installation steps need to be clarified and improved
> on. For example, I tried to use the configure script on my computer
> (Mac) but it didn't seem to work. So I resorted to this install command,
> 
> R CMD INSTALL h5 --configure-vars="'LIBS=-L/usr/local/Cellar/hdf5/1.10.0-patch1 -L. -lhdf5_cpp -lhdf5 -lz
> -lm' CPPFLAGS='-I/usr/local/include'"
> 
> which worked, but it required that I set all the libraries manually, and
> it should be done internally via |src/Makevars|, and allow me to set
> additional flags (e.g., paths to libraries) via |~/.R/Makevars| or by
> passing options to |R CMD INSTALL|.
It seems that the root cause of your problem is the installed Anaconda 3
and a modified path variable which leads to the linking problem (had the
same issue). I understand that h5 should automatically link against any
pre-installed hdf5 library but that is not as easy as it sounds. The
preferred and supported way to install hdf5 is through brew and it is
assumed that the brew path is preferred against any other libray. An
alternative way would be to temporarily disable the Anaconda 3 settings
in .bash_profile.

> Note that the |R CMD INSTALL| command you suggest in the README includes
> a bunch of additional libraries that don't seem necessary, and addition
> you got the quotes wrong.
jup - that seems indeed wrong.

> Also, I have already installed the hdf5 libraries via Anaconda 3, but it
> didn't work when I tried to link to these libraries. It would be helpful
> if you could explain why this doesn't work. (The error message from
> running |R CMD INSTALL| was not helpful.)
see above.

> 
> i hope all these suggestions are useful.
yes - thanks!

Best,
Mario

**Comment by pcarbo** [comment]
Hi Mario, to be clear, I *was* able to install `h5` with Anaconda following the steps I gave above (I ammended my post above).

One final thought: you might want to take a look at how other packages on CRAN have set up compilation and linking to libraries, and try to follow their example---two good examples that come to mind are [sp](http://cran.r-project.org/package=sp) and [proj](http://cran.r-project.org/package=project).

**Comment by mannau** [comment]
Peter,
the configuration file which automatically sets compilation and linking works for most use cases (at least all documented ones) accross all major platforms. I know your suggested packages and the CRAN documentation, but the problem is, that **h5** needs to use `h5c++` to automatically find the include and library paths:
LIBS: https://github.com/mannau/h5/blob/master/configure.ac#L35
CPPFLAGS: https://github.com/mannau/h5/blob/master/configure.ac#L60
Since you somehow messed up your path/`h5c++` with anaconda you may need to use custom install parameters.

For any future suggestions please note that:
- Reporting Problems as Issues is ok, Fixing Problems through Pull Request is highly welcome.
- Reporting Problems and suggesting the package author to read some documentation without really knowing whats going on does not help at all.

-m

**Comment by pcarbo** [comment]
I agree that my comments aren't directly helpful. If I had time I certainly would attempt to correct these issues myself (via a pull request). I am making these suggestions because your R package seems very useful, and potentially of a lot of value to many others. I can tell you that currently the installation process is difficult enough that it is preventing it from being more accessible. I have installed hundreds of R packages on Mac OS X and different operating systems (including high-performance computing setups), and I would say that `h5` is the most difficult package I have ever had to install. Therefore, I made some suggestions for making the instructions more clear. For example, from the README it is not clear whether the `configure` script should be used at all, and whether it works on Mac OS X. (The README says to directly set the `LDFLAGS`, and does not mention the option of using the `configure` script for Mac.) I think these are relatively small issues that could be fixed and make your package more widely used. Good luck.
